[GH-511]: Added workflow to check HTML using HTMLProofer

varunKT001 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary

I have added a workflow to check HTML using HTMLProofer. It uses [Proof HTML](https://github.com/marketplace/actions/proof-html) GitHub Action.
<!--
A description of what this pull request does.
-->

#### Ticket Link

Fixes #511 
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

mattermod commented:
Hello @varunKT001,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

Per the [Mattermost Contribution Guide](https://developers.mattermost.com/contribute/getting-started/), we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**
Once you have signed the CLA, please comment with `/check-cla` and confirm that the CLA check is green.

This is a standard procedure for many open source projects.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

varunKT001 commented:
/check-cla

neflyte commented:
Thanks for your contribution, @varunKT001! 
Since there will be many contributors to the docs, we will need a method to run html-proofer outside of a GitHub action (i.e., as a `Makefile` target).

Please add targets to the `Makefile` to run html-proofer.

neflyte reviewed (changes requested):
Please add `Makefile` targets to run html-proofer locally

varunKT001 commented:
@neflyte I have made the `Makefile` target. Please review.

neflyte commented on Makefile:
```diff
@@ -27,3 +27,7 @@ run:
+
+.PHONY: test
+test:
+	docker run -v $(PWD)/dist/html/:/mnt 18fgsa/html-proofer mnt --ignore_empty_alt
```
We should not need to use docker or any other container to run local testing against the docs. Please add a `Makefile` target to run html-proofer without using a container or docker.

varunKT001 replied:
@neflyte I have updated the Makefile. Please review and suggest if any changes are needed.

varunKT001 commented:
@neflyte I have updated the Makefile. Please review and suggest if any changes are needed.

varunKT001 commented:
@hanzei I'm sorry, I requested @neflyte for a review but somehow github removed you from review 😞

hanzei commented on Makefile:
```diff
@@ -28,9 +28,14 @@ build:
+
+.PHONY: test
+test:
+	htmlproofer ./dist/html --ignore_empty_alt
```
Can you please run that check as part of the CI runs?

varunKT001 replied:
@hanzei I didn't understand 😅 can you please explain a bit?

varunKT001 commented:
@hanzei I didn't understand 😅 Could you please tell me how to use it in CI runs?

neflyte commented on Makefile:
```diff
@@ -28,9 +28,14 @@ build:
+
+.PHONY: test
+test:
+	htmlproofer ./dist/html --ignore_empty_alt
```
This command fails. How do I get the dependencies I need to run the command?

hanzei replied:
Maybe we should make use of https://html-validate.org/usage/cli.html as it's a js tool and therefore dev already installed the tools needed to run it.

varunKT001 commented:
@neflyte @hanzei I have made the required changes. Please review and suggest if any changes are needed.

varunKT001 commented:
@neflyte 

There is no support provided for JavaScript or Go by HTMLProofer. Currently, it's not possible to use HTMLProofer without Ruby. One possible solution is to use Docker.

Usage with Docker: [18fgsa/html-proofer](https://hub.docker.com/r/18fgsa/html-proofer/)

varunKT001 commented:
@neflyte Please suggest if any changes are needed.

varunKT001 commented:
@hanzei @neflyte  

Used `html-validate` instead of `html-proofer` to check generated HTML files locally. Please review the changes.

__Steps to run:__

1. Install `html-validate` globally:

```bash
npm i -g html-validate
```

2. Test the generated HTML files:

```bash
make test
```

neflyte reviewed (changes requested):
We should not require anyone to install global tools.
Since your solution is JS-based, we can move the `package.json` and `package-lock.json` files up one directory and add a dependency for the validation tool.

This way we're only installing the tools we need for a single repo and not for a whole machine.

varunKT001 commented:
@neflyte Made the changes. I added an NPM script instead of the global install 👍

neflyte commented on .gitignore:
```diff
@@ -1,16 +1,19 @@
+/site/data/PluginGoDocs.json
+/site/data/PluginManifestDocs.json
+/site/data/PluginJSDocs.json
+/scripts/node_modules
```
This line can be removed if there will no longer be a `node_modules` subdirectory in `scripts`.

varunKT001 replied:
I have removed it from the `.gitignore` file.

varunKT001 commented on Makefile:
```diff
@@ -28,9 +28,15 @@ build:
+
+.PHONY: test
+test:
+	npm install
```
The build error is because I have created a GitHub Action (@hanzei requested) to build and check the generated HTML files using HTMLProofer. 

I have tested the `test` command. The `npm install` won't cause any issues.

varunKT001 commented:
@neflyte I have checked on both Windows and WSL2, and it works fine. I don't know what the problem is. I have added a separate `package.json` file to the `/scripts` directory. Maybe this should fix the issue.

- Windows:

```bash
$ make dist
mkdir -p site/data
go run ./cmd/plugin-godocs > site/data/PluginGoDocs.json
go run ./cmd/plugin-manifest-docs > site/data/PluginManifestDocs.json
2022/10/27 20:40:04 unrecognized types.Type: *types.Interface, any
2022/10/27 20:40:04 unrecognized types.Type: *types.Interface, any
rm -rf scripts/mattermost-webapp || true
cd scripts && npm install

up to date, audited 13 packages in 1s

3 packages are looking for funding   
  run `npm fund` for details

found 0 vulnerabilities
mkdir -p site/data
node scripts/plugin-jsdocs.js > site/data/PluginJSDocs.json
rm -rf ./dist
hugo -s site --destination ../dist/html
Start building sites … 
hugo v0.104.3-58b824581360148f2d91f5cc83f69bd22c1aa331 windows/amd64 BuildDate=2022-10-04T14:25:23Z VendorInfo=gohugoio

                   | EN   
-------------------+------
  Pages            | 383  
  Paginator pages  |   3  
  Non-page files   | 203  
  Static files     | 156  
  Processed images |   0  
  Aliases          | 183  
  Sitemaps         |   1
  Cleaned          |   0

Total in 8183 ms
```

- Linux (WSL2 - Ubuntu)

```bash
make dist
mkdir -p site/data
go run ./cmd/plugin-godocs > site/data/PluginGoDocs.json
go run ./cmd/plugin-manifest-docs > site/data/PluginManifestDocs.json
2022/10/27 21:28:36 unrecognized types.Type: *types.Interface, any
2022/10/27 21:28:36 unrecognized types.Type: *types.Interface, any
rm -rf scripts/mattermost-webapp || true
cd scripts && npm install

up to date, audited 13 packages in 1s

3 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
mkdir -p site/data
node scripts/plugin-jsdocs.js > site/data/PluginJSDocs.json
rm -rf ./dist
hugo -s site --destination ../dist/html
Start building sites … 
hugo v0.104.3-58b824581360148f2d91f5cc83f69bd22c1aa331 linux/amd64 BuildDate=2022-10-04T14:25:23Z VendorInfo=gohugoio

                   | EN   
-------------------+------
  Pages            | 383  
  Paginator pages  |   3  
  Non-page files   | 203  
  Static files     | 156  
  Processed images |   0  
  Aliases          | 183  
  Sitemaps         |   1  
  Cleaned          |   0  

Total in 20582 ms
```

neflyte commented:
It looks like the package.json issue is fixed. The build is failing for a different reason now.
```
$ make dist
mkdir -p site/data
go run ./cmd/plugin-godocs > site/data/PluginGoDocs.json
go run ./cmd/plugin-manifest-docs > site/data/PluginManifestDocs.json
2022/10/31 10:10:19 unrecognized types.Type: *types.Interface, any
2022/10/31 10:10:19 unrecognized types.Type: *types.Interface, any
rm -rf scripts/mattermost-webapp || true
cd scripts && npm install

added 12 packages, and audited 13 packages in 558ms

3 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
mkdir -p site/data
node scripts/plugin-jsdocs.js > site/data/PluginJSDocs.json
/home/alan/src/varunkt001-mm-dev-docs/scripts/node_modules/espree/lib/espree.js:195
            throw err;
            ^

SyntaxError: Unexpected token :
    at Espree.raise (/home/alan/src/varunkt001-mm-dev-docs/scripts/node_modules/espree/lib/espree.js:190:25)
    at Espree.unexpected (/home/alan/src/varunkt001-mm-dev-docs/scripts/node_modules/espree/lib/espree.js:235:18)
    at Espree.pp.semicolon (/home/alan/src/varunkt001-mm-dev-docs/scripts/node_modules/acorn/dist/acorn.js:674:66)
    at Espree.pp$1.parseExpressionStatement (/home/alan/src/varunkt001-mm-dev-docs/scripts/node_modules/acorn/dist/acorn.js:1153:10)
    at Espree.pp$1.parseStatement (/home/alan/src/varunkt001-mm-dev-docs/scripts/node_modules/acorn/dist/acorn.js:888:26)
    at Espree.pp$1.parseTopLevel (/home/alan/src/varunkt001-mm-dev-docs/scripts/node_modules/acorn/dist/acorn.js:754:23)
    at Espree.parseTopLevel (/home/alan/src/varunkt001-mm-dev-docs/scripts/node_modules/espree/lib/espree.js:178:26)
    at Espree.parse (/home/alan/src/varunkt001-mm-dev-docs/scripts/node_modules/acorn/dist/acorn.js:554:17)
    at Espree.parse (/home/alan/src/varunkt001-mm-dev-docs/scripts/node_modules/espree/lib/espree.js:144:35)
    at Object.parse (/home/alan/src/varunkt001-mm-dev-docs/scripts/node_modules/espree/espree.js:134:38) {
  index: 3,
  lineNumber: 1,
  column: 4
}
make: *** [Makefile:20: frontend-plugin-data] Error 1
```

The GitHub build pipeline is also failing with the same error.

neflyte commented:
It looks like the error above applies to the whole repo. It is not the fault of this PR. :) We're working to fix the build issue.

cwarnermm commented:
@varunKT001 - Are you open to resolving the merge conflicts in this PR?

varunKT001 commented:
@cwarnermm Sure

hanzei commented:
/update-branch

mattermod commented:
Error trying to update the PR.
Please do it manually.

hanzei reviewed (changes requested):
Please merge `master` into your PR to get https://github.com/mattermost/mattermost-developer-documentation/commit/4cc48a07d69d9aaadc79f81175eb80d9e0509481 in

varunKT001 commented:
@hanzei @neflyte @cwarnermm Merged branch `master` onto `varunKT001/check-html`

hanzei commented on .github/workflows/build.yml:
```diff
@@ -30,3 +30,9 @@ jobs:
         run: make dist
+
+      - name: Check HTML
+        uses: anishathalye/proof-html@v1
```
I'm confused. Why is a different tool and not `html -validate` used here?

varunKT001 replied:
@hanzei I've made the changes. Now, the workflow also uses `html-validate`.

varunKT001 commented:
@hanzei I've made the changes. Now, the workflow also uses `html-validate`.

neflyte commented:
@varunKT001 You have build errors due to your change; this is good! The errors are caused by invalid HTML. :slightly_smiling_face: 

Since the `blog` section of the docs is no longer used, please add an exception or exclusion for that section. After making the change, the build should succeed. :smile:

varunKT001 commented:
@neflyte I tried excluding the `/dist/html/blog` directory. However, There are some other directories for which `html-validate` is generating errors. 

`/dist/html/blog`, `/dist/html/integrate`, `/dist/html/internal`, etc. are a few of them.

![image](https://user-images.githubusercontent.com/83509023/200375207-133b7056-1f30-42bc-b5fc-585fe898111a.png)
![image](https://user-images.githubusercontent.com/83509023/200375243-079ae939-d7e5-4c96-93d2-f2355ecd0ead.png)

neflyte commented:
@varunKT001 Would you be willing to fix the existing errors in the generated HTML files?

If not, then we will need to exclude all directories so the build pipeline can succeed. Then we can post additional issues to fix the HTML files, one section at a time.

varunKT001 commented:
@neflyte I have excluded the directories containing the invalid HTML. The build should succeed now.

varunKT001 commented:
@neflyte If the changes look satisfactory, please consider merging the pull request :)

cwarnermm commented:
@neflyte @hanzei - This PR is now ready for your review.

neflyte reviewed (approved):
No errors after your latest change; that's awesome. :smile: 

In the future, we'll need to fix the rest of the build pipelines to run the HTML check, including the production build pipeline.

hanzei commented on Makefile:
```diff
@@ -28,9 +28,15 @@ build:
 	mkdir -p site/static/font
 	curl --no-progress-meter -o site/static/css/compass-icons.css https://mattermost.github.io/compass-icons/css/compass-icons.css
-	curl --no-progress-meter -o "site/static/font/compass-icons.#1" "https://mattermost.github.io/compass-icons/font/compass-icons.{eot,woff2,woff,ttf,svg}"
+	curl --no-progress-meter -o "site/static/font/compass-icons.#1" "https://mattermost.github.io/compass-icons/font/compass-icons.{eot,woff2,woff,ttf,svg}"
```
Missing linebreak

varunKT001 replied:
I can't figure out how to change that 😓
Adding a linebreak will introduce a new line 🤔

hanzei reviewed (approved):
LGTM :tada:

varunKT001 commented:
@cwarnermm Is it ready to be merged now?

hanzei commented:
Merged :tada:
